core/src/main/java/hudson/cli/CLICommand.java

```diff
@@ -253,56 +255,59 @@ public int main(List<String> args, Locale locale, InputStream stdin, PrintStream
-            logFailedCommandAndPrintExceptionErrorMessage(args, e);
-            return 4;
-        } catch (IllegalArgumentException e) {
-            logFailedCommandAndPrintExceptionErrorMessage(args, e);
-            return 3;
-        } catch (AbortException e) {
-            logFailedCommandAndPrintExceptionErrorMessage(args, e);
-            return 5;
-        } catch (AccessDeniedException e) {
-            logFailedCommandAndPrintExceptionErrorMessage(args, e);
-            return 6;
-        } catch (BadCredentialsException e) {
-            // to the caller, we can't reveal whether the user didn't exist or the password didn't match.
-            // do that to the server log instead
-            String id = UUID.randomUUID().toString();
-            logAndPrintError(e, "Bad Credentials. Search the server log for " + id + " for more details.",
-                    "CLI login attempt failed: " + id, Level.INFO);
-            return 7;
         } catch (Throwable e) {
-            String errorMsg = "Unexpected exception occurred while performing " + getName() + " command.";
-            logAndPrintError(e, errorMsg, errorMsg, Level.WARNING);
-            Functions.printStackTrace(e, stderr);
-            return 1;
+            int exitCode = handleException(e, correlationId, p);
+            CliListener.fireError(correlationId, getName(), argsSize, getTransportAuthentication2(), exitCode, e);
+            return exitCode;
         } finally {
             if (sc != null)
                 sc.setAuthentication(old); // restore
         }
     }
 
-    private void logFailedCommandAndPrintExceptionErrorMessage(List<String> args, Throwable e) {
-        Authentication auth = getTransportAuthentication2();
-        String logMessage = String.format("Failed call to CLI command %s, with %d arguments, as user %s.",
-                getName(), args.size(), auth != null ? auth.getName() : "<unknown>");
-
-        logAndPrintError(e, e.getMessage(), logMessage, Level.FINE);
+    /**
+     * Determines command stderr output and return the exit code as described on {@link #main(List, Locale, InputStream, PrintStream, PrintStream)}
+     * */
+    protected int handleException(Throwable e, String correlationId, CmdLineParser p) {
+        int exitCode;
+        if (e instanceof CmdLineException) {
+            exitCode = 2;
+            printError(e.getMessage());
+            printUsage(stderr, p);
+        } else if (e instanceof IllegalArgumentException) {
+            exitCode = 3;
+            printError(e.getMessage());
+        } else if (e instanceof IllegalStateException) {
+            exitCode = 4;
+            printError(e.getMessage());
+        } else if (e instanceof AbortException) {
+            exitCode = 5;
+            printError(e.getMessage());
+        } else if (e instanceof AccessDeniedException) {
+            exitCode = 6;
+            printError(e.getMessage());
+        } else if (e instanceof BadCredentialsException) {
```

Comment:
Looks like `BadCredentialsException` is not longer possible inside `CLICommand#main`, note `{set/get}TransportAuth2`:

* for HTTP or WebSocket the BadCredential is thrown/handled in the filters. When using command line `-auth user:pass` basic auth : BasicHeaderAuthenticator / AbstractUserAuth.

* for SSH auth errors (not a BadCredentials) is also handled before executing the command (PublicKeyAuthenticatorImpl / UserAuthNamedFactory).

https://github.com/jenkinsci/jenkins/commit/b1803a9539ec773e91a9fd126a7ae3c17c98794b I can't find CliAuthenticationTest, may not apply anymore.